File: src/embedding_service.py
"""
EmbeddingService — async Ollama client for text embeddings.

Provides vector embeddings for hybrid (BM25 + pgvector) search across food,
health supplements, and drug modules. Falls back to None on any error so callers
can degrade gracefully to keyword-only search.

Config env vars:
  OLLAMA_BASE_URL          e.g. http://192.168.1.100:11434  (leave unset to disable)
  OLLAMA_EMBED_MODEL       default: qwen3-embedding:0.6b
  OLLAMA_EMBED_DIMENSIONS  output vector size, default: 1024
  OLLAMA_EMBED_TIMEOUT     seconds, default: 30
  OLLAMA_EMBED_BATCH_SIZE  default: 32
"""

from __future__ import annotations

import os
import time
from typing import Sequence

import httpx

import metrics as _metrics
from utils import log_info, log_warning

_BASE_URL: str = os.getenv("OLLAMA_BASE_URL", "").rstrip("/")
_MODEL: str = os.getenv("OLLAMA_EMBED_MODEL", "qwen3-embedding:0.6b")
_TIMEOUT: float = float(os.getenv("OLLAMA_EMBED_TIMEOUT", "30"))
BATCH_SIZE: int = int(os.getenv("OLLAMA_EMBED_BATCH_SIZE", "32"))
DIMENSIONS: int = int(os.getenv("OLLAMA_EMBED_DIMENSIONS", "1024"))


def configure(values: dict) -> None:
    """Override the Ollama embedding settings from a DB settings dict
    (admin_settings 'embedding' group). Called at startup and on settings save
    so query-time semantic search picks up changes without a restart."""
    global _BASE_URL, _MODEL, _TIMEOUT, BATCH_SIZE, DIMENSIONS, _PROVIDER, _API_KEY
    _PROVIDER = str(values.get("provider", "ollama") or "ollama").strip().lower()
    _BASE_URL = str(values.get("base_url", "") or "").rstrip("/")
    _API_KEY = str(values.get("api_key", "") or "")
    _MODEL = str(values.get("model", "") or "")
    _TIMEOUT = float(values.get("timeout", 30) or 30)
    BATCH_SIZE = int(values.get("batch_size", 32) or 32)
    DIMENSIONS = int(values.get("dimensions", 1024) or 1024)


_PROVIDER: str = "ollama"
_API_KEY: str = ""
_GOOGLE_BASE = "https://generativelanguage.googleapis.com"


def _provider_operation(operation: str) -> str:
    return f"{_PROVIDER}.{operation}"


def _record_embedding_call(operation: str, status: str, start: float) -> None:
    _metrics.record_dependency_call(
        "embedding",
        _provider_operation(operation),
        status,
        time.monotonic() - start,
    )


def _configured() -> bool:
    """Whether the embedding provider has enough config to attempt a call."""
    if _PROVIDER == "ollama":
        return bool(_BASE_URL)
    return bool(_MODEL and _API_KEY)


async def _provider_embed(client, texts: list[str]) -> list:
    """Embed a batch via the configured provider. Raises on HTTP error."""
    if _PROVIDER == "openai":
        base = (
            _BASE_URL
            if _BASE_URL.endswith("/embeddings")
            else f"{_BASE_URL}/embeddings"
        )
        body: dict = {"model": _MODEL, "input": texts, "encoding_format": "float"}
        if DIMENSIONS and _MODEL.startswith("text-embedding-3"):
            body["dimensions"] = DIMENSIONS
        headers = {"Authorization": f"Bearer {_API_KEY}"} if _API_KEY else {}
        resp = await client.post(base, json=body, headers=headers)
        resp.raise_for_status()
        data = sorted(resp.json().get("data", []), key=lambda d: d.get("index", 0))
        return [d.get("embedding") for d in data]
    if _PROVIDER == "google":
        model_path = _MODEL if _MODEL.startswith("models/") else f"models/{_MODEL}"
        reqs = []
        for t in texts:
            req = {"model": model_path, "content": {"parts": [{"text": t}]}}
            if DIMENSIONS:
                req["outputDimensionality"] = DIMENSIONS
            reqs.append(req)
        resp = await client.post(
            f"{_GOOGLE_BASE}/v1beta/{model_path}:batchEmbedContents",
            json={"requests": reqs},
            headers={"x-goog-api-key": _API_KEY},
        )
        resp.raise_for_status()
        return [e.get("values") for e in resp.json().get("embeddings", [])]
    resp = await client.post(
        f"{_BASE_URL}/api/embed", json={"model": _MODEL, "input": texts}
    )
    resp.raise_for_status()
    return resp.json().get("embeddings", [])


class EmbeddingService:
    """Async Ollama embedding client with automatic fallback to keyword-only search."""

    def __init__(self) -> None:
        self._available: bool = False

    async def initialize(self) -> None:
        if not _configured():
            log_warning(
                "Embedding provider not configured — semantic search disabled, using keyword-only"
            )
            return
        self._available = await self._ping()
        if self._available:
            log_info(
                "EmbeddingService ready",
                provider=_PROVIDER,
                model=_MODEL,
                base_url=_BASE_URL,
            )
        else:
            log_warning(
                "Embedding provider not reachable at startup — will retry on first query",
                provider=_PROVIDER,
                base_url=_BASE_URL,
            )

    async def _ping(self) -> bool:
        if not _configured():
            return False
        start = time.monotonic()
        try:
            if _PROVIDER == "ollama":
                async with httpx.AsyncClient(timeout=3.0) as client:
                    r = await client.get(f"{_BASE_URL}/api/version")
                    ok = r.status_code == 200
                    _record_embedding_call(
                        "ping",
                        "success" if ok else "error",
                        start,
                    )
                    return ok
            # openai / google: a successful 1-item embed is the cheapest real check
            async with httpx.AsyncClient(timeout=8.0) as client:
                out = await _provider_embed(client, ["ping"])
                ok = bool(out and out[0])
                _record_embedding_call("ping", "success" if ok else "error", start)
                return ok
        except Exception:
            _record_embedding_call("ping", "error", start)
            return False

    @property
    def enabled(self) -> bool:
        return _configured()

    async def embed(self, text: str) -> list[float] | None:
        """Embed a single text string via Ollama.

        Args:
            text: The input text to embed.

        Returns:
            A float vector, or ``None`` if Ollama is unavailable.
        """
        results = await self.embed_batch([text])
        return results[0]

    async def embed_batch(self, texts: Sequence[str]) -> list[list[float] | None]:
        """Embed multiple texts in one Ollama call.

        Returns a list parallel to *texts*; each item is either a float list
        (success) or None (failure for that item).
        """
        if not _configured() or not texts:
            return [None] * len(texts)
        start = time.monotonic()
        try:
            async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
                embeddings = await _provider_embed(client, list(texts))
            _record_embedding_call("embed_batch", "success", start)
            result: list[list[float] | None] = []
            for i in range(len(texts)):
                result.append(embeddings[i] if i < len(embeddings) else None)
            if not self._available:
                log_info("Embedding connection restored — semantic search re-enabled")
            self._available = True
            return result
        except Exception as exc:
            _record_embedding_call("embed_batch", "error", start)
            if self._available:
                log_warning(
                    "Embedding failed — falling back to keyword search",
                    error=str(exc),
                )
            self._available = False
            return [None] * len(texts)

File: node-server/src/embeddingService.ts
/**
 * EmbeddingService — async embedding client (Ollama / OpenAI / Google).
 *
 * Faithful port of `src/embedding_service.py`. Provides vector embeddings for
 * hybrid (BM25 + pgvector) search. Falls back to `null` on any error so callers
 * degrade gracefully to keyword-only search.
 *
 * Config is **not** read from env here (those are seed-only in this project).
 * Settings live in `admin.app_settings` group `embedding`, exactly as the Python
 * lifespan applies them via `embedding_service.configure(...)`. `getEmbeddingService`
 * loads that group from the DB once and caches the configured singleton.
 */

import { query } from "./db.js";
import { logInfo, logWarning } from "./logger.js";

export interface EmbeddingSettings {
  provider: string;
  baseUrl: string;
  apiKey: string;
  model: string;
  timeout: number;
  batchSize: number;
  dimensions: number;
}

const GOOGLE_BASE = "https://generativelanguage.googleapis.com";

function configured(s: EmbeddingSettings): boolean {
  if (s.provider === "ollama") return Boolean(s.baseUrl);
  return Boolean(s.model && s.apiKey);
}

/** Embed a batch via the configured provider. Throws on HTTP error. */
async function providerEmbed(s: EmbeddingSettings, texts: string[]): Promise<(number[] | null)[]> {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), s.timeout * 1000);
  try {
    if (s.provider === "openai") {
      const base = s.baseUrl.endsWith("/embeddings") ? s.baseUrl : `${s.baseUrl}/embeddings`;
      const body: Record<string, unknown> = {
        model: s.model,
        input: texts,
        encoding_format: "float",
      };
      if (s.dimensions && s.model.startsWith("text-embedding-3")) {
        body.dimensions = s.dimensions;
      }
      const headers: Record<string, string> = { "Content-Type": "application/json" };
      if (s.apiKey) headers.Authorization = `Bearer ${s.apiKey}`;
      const resp = await fetch(base, { method: "POST", headers, body: JSON.stringify(body), signal: ctrl.signal });
      if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
      const json = (await resp.json()) as { data?: { index?: number; embedding?: number[] }[] };
      const data = [...(json.data ?? [])].sort((a, b) => (a.index ?? 0) - (b.index ?? 0));
      return data.map((d) => d.embedding ?? null);
    }
    if (s.provider === "google") {
      const modelPath = s.model.startsWith("models/") ? s.model : `models/${s.model}`;
      const reqs = texts.map((t) => {
        const req: Record<string, unknown> = { model: modelPath, content: { parts: [{ text: t }] } };
        if (s.dimensions) req.outputDimensionality = s.dimensions;
        return req;
      });
      const resp = await fetch(`${GOOGLE_BASE}/v1beta/${modelPath}:batchEmbedContents`, {
        method: "POST",
        headers: { "Content-Type": "application/json", "x-goog-api-key": s.apiKey },
        body: JSON.stringify({ requests: reqs }),
        signal: ctrl.signal,
      });
      if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
      const json = (await resp.json()) as { embeddings?: { values?: number[] }[] };
      return (json.embeddings ?? []).map((e) => e.values ?? null);
    }
    // ollama
    const resp = await fetch(`${s.baseUrl}/api/embed`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ model: s.model, input: texts }),
      signal: ctrl.signal,
    });
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    const json = (await resp.json()) as { embeddings?: number[][] };
    return json.embeddings ?? [];
  } finally {
    clearTimeout(timer);
  }
}

export class EmbeddingService {
  private settings: EmbeddingSettings;
  private available_ = false;

  constructor(settings: EmbeddingSettings) {
    this.settings = settings;
  }

  get enabled(): boolean {
    return configured(this.settings);
  }

  /**
   * Reachability flag — mirrors Python `embedding_service._available`: false
   * until a successful embed (or init ping), flipped false on failure. Used by
   * `check_embedding_health` to decide semantic vs keyword-only.
   */
  get available(): boolean {
    return this.available_;
  }

  /**
   * Mirror Python `initialize()` → `_ping()`: probe the provider once with a
   * trivial embed so `available` reflects reachability before any real query.
   * Fail-open (the embed catch already records availability).
   */
  async initialize(): Promise<void> {
    if (!configured(this.settings)) return;
    await this.embed("ping");
  }

  /** Embed a single text string. Returns the vector, or null when unavailable. */
  async embed(text: string): Promise<number[] | null> {
    const results = await this.embedBatch([text]);
    return results[0] ?? null;
  }

  /** Embed multiple texts in one call. Each item is a vector or null. */
  async embedBatch(texts: string[]): Promise<(number[] | null)[]> {
    if (!configured(this.settings) || texts.length === 0) {
      return texts.map(() => null);
    }
    try {
      const embeddings = await providerEmbed(this.settings, texts);
      const result: (number[] | null)[] = [];
      for (let i = 0; i < texts.length; i++) {
        result.push(i < embeddings.length ? embeddings[i] : null);
      }
      if (!this.available_) {
        logInfo("Embedding connection restored — semantic search re-enabled");
      }
      this.available_ = true;
      return result;
    } catch (exc) {
      if (this.available_) {
        logWarning("Embedding failed — falling back to keyword search", {
          error: String((exc as Error).message),
        });
      }
      this.available_ = false;
      return texts.map(() => null);
    }
  }
}

/** Load the `embedding` settings group from admin.app_settings. Fails open. */
export async function loadEmbeddingSettings(): Promise<EmbeddingSettings> {
  const defaults: EmbeddingSettings = {
    provider: "ollama",
    baseUrl: "",
    apiKey: "",
    model: "",
    timeout: 30,
    batchSize: 32,
    dimensions: 1024,
  };
  try {
    const res = await query<{ key: string; value: string | null }>(
      "SELECT key, value FROM admin.app_settings WHERE group_key = 'embedding'",
    );
    const m = new Map(res.rows.map((r) => [r.key, r.value ?? ""]));
    return {
      provider: (m.get("provider") || "ollama").trim().toLowerCase(),
      baseUrl: (m.get("base_url") || "").replace(/\/+$/, ""),
      apiKey: m.get("api_key") || "",
      model: m.get("model") || "",
      timeout: Number(m.get("timeout")) || 30,
      batchSize: Number(m.get("batch_size")) || 32,
      dimensions: Number(m.get("dimensions")) || 1024,
    };
  } catch {
    return defaults;
  }
}

let _svc: EmbeddingService | null = null;

/**
 * Process-wide embedding service singleton. Loads settings from the DB once
 * (mirrors the Python lifespan applying `admin_settings.get_group("embedding")`).
 */
export async function getEmbeddingService(): Promise<EmbeddingService> {
  if (_svc) return _svc;
  const settings = await loadEmbeddingSettings();
  _svc = new EmbeddingService(settings);
  if (settings.provider && configured(settings)) {
    // Mirror Python lifespan: ping on init so `available` reflects reachability.
    await _svc.initialize();
    logInfo("EmbeddingService ready", {
      provider: settings.provider,
      model: settings.model,
      base_url: settings.baseUrl,
    });
  } else {
    logWarning("Embedding provider not configured — semantic search disabled, using keyword-only");
  }
  return _svc;
}

/**
 * Drop the cached singleton so the next `getEmbeddingService()` reloads settings
 * from `admin.app_settings` and re-pings. Mirrors the Python
 * `_refresh_settings_singletons` path that calls `embedding_service.configure(...)`
 * after an `embedding` settings save — no process restart needed.
 */
export async function reconfigureEmbeddingService(): Promise<void> {
  _svc = null;
  await getEmbeddingService();
}

/** Format a vector as a pgvector/halfvec literal `[a,b,c]`, or null. */
export function vecLiteral(vec: number[] | null): string | null {
  if (!vec || vec.length === 0) return null;
  return `[${vec.join(",")}]`;
}
